production_entry_app/production_entry_app/joint_production.py

```
@@ -0,0 +1,572 @@
+			)
+
+	# Native Stock Entry recalculation applies the manual rates consistently to
+	# basic amounts, valuation rates, totals, and the eventual ledger entries.
+	doc.calculate_rate_and_amount(reset_outgoing_rate=False)
+
+
+def _is_scrap_row(row: Any) -> bool:
+	return bool(row.get("is_scrap_item") or row.get("is_legacy_scrap_item") or row.get("type") == "Scrap")
+
+
+def _validate_joint_rejection_breakup(doc: Any, plan: JointProductionPlan) -> None:
```

Ruff (0.16.1)

[warning] Dynamically typed expressions (typing.Any) are disallowed in `doc`

(ANN401)